Adding CUE Advanced EN (#245)
oam-dev/kubevela.io: docs/platform-engineers/cue/advanced.md

title: CUE Advanced
---

> This section will introduce how to use CUE to deliver complete modular functions, so that your platform can dynamically expand functions as user needs change, adapt to various users and scenarios, and meet the iterative demands of the company's long-term business development.

Reviewer: ```suggestion
This section will introduce how to use CUE to deliver KubeVela modules. You can can dynamically expand the platform as user needs change, adapt to growing number of users and scenarios, and meet the iterative demands of the company's long-term business development.
```

## Convert Kubernetes API Objects Into Custom Components

> Let's take the official [StatefulSet object][5] of Kubernetes as an example to see how to use KubeVela to build custom modular functions and provide capabilities.

Reviewer: ```suggestion
Let's take the [Kubernetes StatefulSet][5] as an example to show how to use KubeVela to build custom modules and provide capabilities.
```

> We save the YAML example of StatefulSet in the official document locally and name it as `my-stateful.yaml`, then execute commande as below:

Reviewer: ```suggestion
Save the YAML example of StatefulSet in the official document locally and name it as `my-stateful.yaml`, then execute commande as below:
```

Reviewer: 尽量避免用 We。就算要用人称也是用 You

	 vela def init my-stateful -t component --desc "My StatefulSet component." --template-yaml ./my-stateful.yaml -o my-stateful.cue

View the generated "my-stateful.cue" file:

	$ cat my-stateful.cue
	"my-stateful": {
		annotations: {}
		attributes: workload: definition: {
			apiVersion: "<change me> apps/v1"
			kind:       "<change me> Deployment"
		}
		description: "My StatefulSet component."
		labels: {}
		type: "component"
	}
	
	template: {
		output: {
			apiVersion: "v1"
			kind:       "Service"
				... // omit non-critical info
		}
		outputs: web: {
			apiVersion: "apps/v1"
			kind:       "StatefulSet"
				... // omit non-critical info
		}
		parameter: {}
	}

> Let's do some fine-tuning of this automatically generated custom component:

Reviewer: ```suggestion
Modify the generated file as follows:
```

		labels: {}
		type: "component"
	}
	
	template: {
		output: {
			apiVersion: "apps/v1"
			kind:       "StatefulSet"
			metadata: name: "web"
			spec: {
				selector: matchLabels: app: "nginx"
				replicas:    3
				serviceName: "nginx"
				template: {
					metadata: labels: app: "nginx"
					spec: {
						containers: [{
							name: "nginx"
							ports: [{
								name:          "web"
								containerPort: 80
							}]
							image: "k8s.gcr.io/nginx-slim:0.8"
							volumeMounts: [{
								name:      "www"
								mountPath: "/usr/share/nginx/html"
							}]
						}]
						terminationGracePeriodSeconds: 10
					}
				}
				volumeClaimTemplates: [{
					metadata: name: "www"
					spec: {
						accessModes: ["ReadWriteOnce"]
						resources: requests: storage: "1Gi"
						storageClassName: "my-storage-class"
					}
				}]
			}
		}
		outputs: web: {
			apiVersion: "v1"
			kind:       "Service"
			metadata: {
				name: "nginx"
				labels: app: "nginx"
			}
			spec: {
				clusterIP: "None"
				ports: [{
					name: "web"
					port: 80
				}]
				selector: app: "nginx"
			}
		}
		parameter: {}
	}

Install ComponentDefinition into the Kubernetes cluster:

> 	$ vela def apply my-stateful.cue

Reviewer: 空行

Reviewer: 未解决 :)

Author: 已更新。刚 commit

	ComponentDefinition my-stateful created in namespace vela-system.

> At this point, the end user of the platform can already see that a `my-stateful` component is available through the `vela components` command.

Reviewer: ```suggestion
You can see that a `my-stateful` component  via `vela components` command:
```

Reviewer: Ping

	$ vela components
	NAME       	NAMESPACE  	WORKLOAD                             	DESCRIPTION
	...
	my-stateful	vela-system	statefulsets.apps                    	My StatefulSet component.
	... 

> By publishing the `Application` of KubeVela to the cluster, we can pull up the StatefulSet and Service objects we just defined.

Reviewer: ```suggestion
Apply the   KubeVela `Application` to the cluster:
```

Reviewer: ping

	cat <<EOF | kubectl apply -f -
	apiVersion: core.oam.dev/v1beta1
	kind: Application
	metadata:
	  name: website
	spec:
	  components:
	    - name: my-component
	      type: my-stateful
	EOF

> 

Reviewer: 这里需要讲一下上面 apply 之后的效果

Reviewer: ping

Author: 具体指要突出什么效果？

这里自定义组件讲解完毕，让读者知道如何用到 YAML 里就结束了，感觉是合理的。

Reviewer: 到这里非常 confusing。为什么要 apply 这个 app? 不是为了说明前面那个 compDef 能用吗？这里不用说明一下吗？

## Define Customized Parameters For Component

> In order to meet the changing needs of users, we need to expose some parameters in the last `parameter`. You can learn about the syntax of parameters in [CUE Basic][7].

Reviewer: > In order to meet the changing needs of users,

没 get 到这个解释的点在哪？有啥逻辑？跟 expose parameters 有啥关系？如果说不清楚干脆别写这段话。

Reviewer: ping

Author: 逻辑是在第一步教如何自定义组件的基础上， 新增了 name、replicas、image  这三个暴露出来的参数。因为是用户有需要才暴露的这三个。如果有用户其它需要，可以暴露其它的参数。

In this example, we expose the following parameters to the user:

* Image name, allowing users to customize the image
* Instance name, allowing users to customize the instance name of the generated StatefulSet object and Service object

> * The number of copies, the number of copies of the generated object

Reviewer: copies -> replica

Reviewer: ping

		... # Omit other unmodified fields
		template: {
			output: {
				apiVersion: "apps/v1"
				kind:       "StatefulSet"
				metadata: name: parameter.name
				spec: {
					selector: matchLabels: app: "nginx"
					replicas:    parameter.replicas
					serviceName: "nginx"
					template: {
						metadata: labels: app: "nginx"
						spec: {
							containers: [{
								image: parameter.image
			
							    ... // Omit other unmodified fields
							}]
						}
					}
					    ... // Omit other unmodified fields
				}
			}
			outputs: web: {
				apiVersion: "v1"
				kind:       "Service"
				metadata: {
					name: "nginx"
					labels: app: "nginx"
				}
				spec: {
					... // Omit other unmodified fields	
			    }
			}
			parameter: {
				image: string
				name: string
				replicas: int
			}
		}

After modification, use `vela def apply` to install to the cluster:

	$ vela def apply my-stateful.cue
	ComponentDefinition my-stateful in namespace vela-system updated.

> This modification process takes effect in real time, and the user can immediately see that the my-stateful component in the system has added new parameters.

Reviewer: ```suggestion
You can see the parameters of my-stateful ComponentDefinition as follows:
```

Reviewer: ping

	$ vela show my-stateful
	# Properties
	+----------+-------------+--------+----------+---------+
	|   NAME   | DESCRIPTION |  TYPE  | REQUIRED | DEFAULT |
	+----------+-------------+--------+----------+---------+
	| name     |             | string | true     |         |
	| replicas |             | int    | true     |         |
	| image    |             | string | true     |         |
	+----------+-------------+--------+----------+---------+

> The modification of the ComponentDefinition will not affect the application that is already running. The new ComponentDefinition will take effect when the application is modified and redeployed next time.

Reviewer: ```suggestion
Updating the ComponentDefinition will not affect existing Applications. It will take effect only after updating the Applications next time.
```

> The end user can specify these three new parameters in the application:

Reviewer: ```suggestion
You can specify the three new parameters in the application:
```

```
	apiVersion: core.oam.dev/v1beta1
	kind: Application
	metadata:
	  name: website
	spec:
	  components:
	    - name: my-component
	      type: my-stateful
	      properties:
	        image: nginx:latest
	        replicas: 1
	        name: my-component
```

Save the file locally and name it `app-stateful.yaml`, execute `kubectl apply -f app-stateful.yaml` to update the application, you can see that the name, image, and number of instances of the StatefulSet object have been updated.

> ## Debugging Modular Functions

Reviewer: ```suggestion
## Dry-run
```

spec:
  clusterIP: None
  ports:
  - name: web
    port: 80
  selector:
    app: nginx
  template:
    spec:
      containers:
      - image: saravak/fluentd:elastic
        name: my-sidecar

---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.oam.dev/appRevision: ""
    app.oam.dev/component: my-component
    app.oam.dev/name: website
    trait.oam.dev/resource: web
    trait.oam.dev/type: AuxiliaryWorkload
  name: web
  namespace: default
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  serviceName: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: k8s.gcr.io/nginx-slim:0.8
        name: nginx
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - mountPath: /usr/share/nginx/html
          name: www
      terminationGracePeriodSeconds: 10
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
      storageClassName: my-storage-class
```

</details>

You can also use `vela dry-run -h` to view more available function parameters.

> ## Use `contxt` to Reduce Parameters

Reviewer: ```suggestion
## Use `context` to get runtime information
```

In our Application example above, the name field in the properties and the name field of the Component are the same. So we can use the `context` keyword that carries context information in the template, where `context.name` is the runtime component Name, thus the name parameter in `parameter` is no longer needed.

	... # Omit other unmodified fields
	template: {
		output: {
			apiVersion: "apps/v1"
			kind:       "StatefulSet"
			metadata: name: context.name
				... // 省略其他没有修改的字段
		}
	    parameter: {
			image: string
			replicas: int
		}
	}

KubeVela has built-in application [required context][9], you can configure it according to your needs.

## Add Traits On Demand

> For new user needs, in addition to modifying ComponentDefinitions and adding parameters, you can also use the TraitDefinition to add configurations as needed. On the one hand, KubeVela has built-in a large number of general operation and maintenance capabilities, which can meet the needs such as: adding labels, annotations, injecting environment variables, sidecars, adding volumes, and so on. On the other hand, like custom component does, you can [Customize Trait ][10] to meet the needs of more configuration and flexible assembly.

Reviewer: ```suggestion
n addition to modifying ComponentDefinitions and adding parameters, you can also use the TraitDefinition to patch configurations to Components.  KubeVela has built-in operations to meet the following needs: adding labels, annotations, injecting environment variables, sidecars, adding volumes, and so on.  You can also [customize Trait][10] to do more flexible patching.
```

	      properties:
	        image: nginx:latest
	        replicas: 1
	        name: my-component
	      traits:
	      - type: sidecar
	        properties:
	          name: my-sidecar
	          image: saravak/fluentd:elastic

Deploy and run the application, and you can see that a fluentd sidecar has been deployed and running in the StatefulSet.

You can also use `vela def` to get the CUE source file of the sidecar to modify, add parameters, etc.

	vela def get sidecar

The customization of operation and maintenance capabilities is similar to component customization, so we won’t go into details here. You can read [Customize Trait][11] for more detailed functions.

## Summarize

This section introduces how to deliver complete modular capabilities through CUE. The core is that it can dynamically increase configuration capabilities according to user needs, and gradually expose more functions and usages, so as to reduce the overall learning threshold for users and ultimately improve R&D efficient.
The out-of-the-box capabilities provided by KubeVela, including components, traits, policy, and workflow, are also designed as pluggable and modifiable capabilities.

## Next

> * Get to know about [customize component](../components/custom-component)

Reviewer: 这种就不要重复前面的内容，而是这样：

get to know about how to customize:
- Component
- Workflow

Reviewer: ping

* Get to know about [customize trait](../traits/customize-trait)
* Get to know about [customize policy](../policy/custom-policy)
* Get to know about [customize workflow](../workflow/custom-workflow)

[1]:	../oam/oam-model
[2]:	../oam/x-definition
[3]:	../cue/basic
[4]:	../cue/definition-edit
[5]:	https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/
[6]:	../components/custom-component#%E4%BA%A4%E4%BB%98%E4%B8%80%E4%B8%AA%E5%A4%8D%E5%90%88%E7%9A%84%E8%87%AA%E5%AE%9A%E4%B9%89%E7%BB%84%E4%BB%B6
[7]:	../cue/basic#%E5%AE%9A%E4%B9%89%E4%B8%80%E4%B8%AA-cue-%E6%A8%A1%E6%9D%BF

> [9]:	../oam/x-definition#%E6%A8%A1%E5%9D%97%E5%AE%9A%E4%B9%89%E8%BF%90%E8%A1%8C%E6%97%B6%E4%B8%8A%E4%B8%8B%E6%96%87

Reviewer 2: 这三个链接看看要不要改成英文的

Author: 要的。已更新成对应英文链接。

							ports: [{
								name:          "web"
								containerPort: 80
							}]
							image: "k8s.gcr.io/nginx-slim:0.8"
							volumeMounts: [{
								name:      "www"
								mountPath: "/usr/share/nginx/html"
							}]
						}]
						terminationGracePeriodSeconds: 10
					}
				}
				volumeClaimTemplates: [{
					metadata: name: "www"
					spec: {
						accessModes: ["ReadWriteOnce"]
						resources: requests: storage: "1Gi"
						storageClassName: "my-storage-class"
					}
				}]
			}
		}
		outputs: web: {
			apiVersion: "v1"
			kind:       "Service"
			metadata: {
				name: "nginx"
				labels: app: "nginx"
			}
			spec: {
				clusterIP: "None"
				ports: [{
					name: "web"
					port: 80
				}]
				selector: app: "nginx"
			}
		}
		parameter: {}
	}

Install ComponentDefinition into the Kubernetes cluster:

	$ vela def apply my-stateful.cue
	ComponentDefinition my-stateful created in namespace vela-system.

You can see that a `my-stateful` component  via `vela components` command:

	$ vela components
	NAME       	NAMESPACE  	WORKLOAD                             	DESCRIPTION
	...
	my-stateful	vela-system	statefulsets.apps                    	My StatefulSet component.
	... 

> Apply the   KubeVela `Application` to the cluster:

Reviewer: ```suggestion
Apply the KubeVela `Application` to the cluster:
```

	cat <<EOF | kubectl apply -f -
	apiVersion: core.oam.dev/v1beta1
	kind: Application
	metadata:
	  name: website
	spec:
	  components:
	    - name: my-component
	      type: my-stateful
	EOF

## Define Customized Parameters For Component

> In order to meet the changing needs of users, we need to expose some parameters in the last `parameter`. You can learn about the syntax of parameters in [CUE Basic][7].

Reviewer: ```suggestion
In previous section we have defined a ComponentDefinition that has no parameter. In this section we will show how to expose parameters.
```

Author: 英文改成这么表达确实贴切很多了。谢谢超哥。
